### PR TITLE
HAAR-819: Added group endpoints ('/api/groups/{group}' and '/api/grou…

### PR DESCRIPTION
**Pull request by ravmoj**
Added group endpoints ('/api/groups/{group}' and '/api/groups/child/{group}')

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+@Schema(description = "Group Name")
+data class GroupAmendment(
+  @Schema(required = true, description = "Group Name", example = "HDC NPS North East")
+  @field:NotBlank(message = "parent group code must be supplied")
+  val groupName: String
+)
+
+@Schema(description = "User Role")
+data class AuthUserAssignableRole(
+  @Schema(required = true, description = "Role Code", example = "LICENCE_RO")
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt:127]: I am assuming we will change the name of this class at some point as it should no longer refer to Auth

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+) {
+
+  constructor(a: Authority, automatic: Boolean) : this(a.roleCode, a.roleName, automatic)
+}
+
+@Schema(description = "User Role")
+data class Authority(
+  @Schema(required = true, description = "Role Code", example = "LICENCE_RO")
+  val roleCode: String,
+
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt:144]: This class is the same as "Role" - we should probably just use one of them

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt [hunk 3]
```diff
@@ -116,6 +118,25 @@ class ExternalUsersApiService(
...
   }
+
+  fun getGroupDetail(group: String): GroupDetails =
+    externalUsersWebClient.get()
+      .uri("/groups/$group")
+      .retrieve()
+      .bodyToMono(GroupDetails::class.java)
+      .block(timeout) ?: throw GroupNotFoundException("get", group, "notfound")
+
+  @Throws(RoleNotFoundException::class)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt:129]: Can remove thrown exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt [hunk 4]
```diff
@@ -116,6 +118,25 @@ class ExternalUsersApiService(
       .toBodilessEntity()
       .block(timeout)
   }
+
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt:121]: Need @Throws ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt [hunk 5]
```diff
@@ -161,6 +162,20 @@ class HmppsManageUsersApiExceptionHandler {
       )
   }
 
+  @ExceptionHandler(GroupNotFoundException::class)
+  fun handleGroupNotFoundException(e: GroupNotFoundException): ResponseEntity<ErrorResponse> {
+    log.debug("Username not found exception caught: {}", e.message)
+    return ResponseEntity
+      .status(NOT_FOUND)
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt:169]: May need to add `.contentType(APPLICATION_JSON)` if you force the exception in postman the type will be wrong

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        .expectBody()
+        .jsonPath("$").value<Map<String, Any>> {
+          assertThat(it["status"] as Int).isEqualTo(NOT_FOUND.value())
+          assertThat(it["userMessage"] as String).startsWith("Group Not found: Unable to maintain group: Not_A_Group with reason: notfound")
+          assertThat(it["developerMessage"] as String).startsWith("Unable to maintain group: Not_A_Group with reason: notfound")
+        }
+    }
+
+    @Test
+    fun `Group details endpoint not accessible without valid token`() {
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt:165]: Again, possibly a missing test
Child Group details when no authority

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt (hunk 2 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt]: now removed this Authority not used in these endpoint...as suggested will use Role from RolesController

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  val roleName: String,
+
+  @Schema(required = true, description = "Role Description", example = "Licence Responsible Officer")
+  val roleDescription: String?,
+
+  @Schema(required = true, description = "Admin type", example = "DPS_LSA")
+  val adminType: List<AdminType> = listOf()
+)
+@Schema(description = "User Group")
+data class AuthUserGroup(
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsController.kt]: Renamed to UserGroup

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt (hunk 4 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/ExternalUsersApiService.kt:121]: Added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt (hunk 5 above)
Comment (ravmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/config/ManageUsersApiExceptionHandler.kt:170]: Added

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    @Test
+    fun `access forbidden when no role`() {
+      webTestClient.get().uri("/groups/SITE_1_GROUP_2")
+        .headers(setAuthorisation(roles = listOf()))
+        .exchange()
+        .expectStatus().isForbidden
+    }
+
+    @Test
+    fun `Group details endpoint returns details of group when user has ROLE_MAINTAIN_OAUTH_USERS`() {
```
[the remark this answers is not in the document]
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt:39]: Added

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt (hunk 6 above)
Comment (ravmoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/GroupsControllerIntTest.kt:211]: Added

**Review by karenmillermoj — APPROVED** [review]
Think we still need to look at whether we should throw exceptions if the call to external users fails - but this is for all calls (not just groups).
Also the ChildGroupNotFoundException doesn't have its associated @Throws declaration - but happy for you to do this in a separate PR

**Pull request merged**